### PR TITLE
Feature/automated soft credits  affiliations

### PR DESCRIPTION
**Pull request by joy-sfdo**
Implemented functionality that allows a data entry user to define which specific Affiliation records should receive OCRs when an Opportunity is created. Work Item #: W-009097.

# Critical Changes

# Changes

# Issues Closed

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 1]
```diff
@@ -383,6 +397,13 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
                     = OPP_AutomatedSoftCreditsService.createRelationshipOCRs(primaryContactsToOpportunities);
                 dmlWrapper.objectsToInsert.addAll(relationshipOCRs);
             }
+
+            // Create Affiliation Soft Credits accordingly
+            if (!accountsToOpportunities.isEmpty()) {
```
Comment (davidhabib) [inline src/classes/OPP_OpportunityContactRoles_TDTM.cls]: same comment as above about trying to optimize our number of soqls against OCR's and avoiding duplicate OCRs.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 2]
```diff
@@ -654,4 +677,34 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
...
+    private static Map<Id, List<Id>> populateAccountToOpportunities(List<Opportunity> currentOpportunities, Map<Id, Account> oppAccountDetails) {
+        Map<Id, List<Id>> accountToOpportunity = new Map<Id, List<Id>>();
+
+        if (currentOpportunities.isEmpty()) {
+            return accountToOpportunity;
+        }
+
+        for (Opportunity eachOppty : currentOpportunities) {
+            if (eachOppty.AccountId != null
+                && !oppAccountDetails.get(eachOppty.AccountId).npe01__SYSTEMIsIndividual__c) {
```
Comment (davidhabib) [inline src/classes/OPP_OpportunityContactRoles_TDTM.cls]: why not use your AutomatedSoftCreditService.isOrganizationalAccount() method (which I still which took an account rather than the flag :-) )

File: src/classes/OPP_AutomatedSoftCreditsService_TEST.cls [hunk 3]
```diff
@@ -242,7 +288,107 @@ private class OPP_AutomatedSoftCreditsService_TEST {
...
+        Test.startTest();
+
+        Contact testCon = [SELECT Id, AccountId FROM Contact WHERE FirstName = 'Affiliation5'][0];
+        Account testAcount = [SELECT Id, npe01__SYSTEMIsIndividual__c FROM Account WHERE Id = :testCon.AccountId][0];
+        System.assertEquals(false, testAcount.npe01__SYSTEMIsIndividual__c);
+        List<OpportunityContactRole> retrievedOCR
+            = [SELECT Id, ContactId, Role
+                FROM OpportunityContactRole
+                WHERE ContactId = :testCon.Id
+                AND Role = :RELATIONSHIP_OPPORTUNITY_CONTACT_ROLE_SOFT_CREDIT];
```
[the remark this answers is not in the document]
Comment (joy-sfdo) [reply src/classes/OPP_AutomatedSoftCreditsService_TEST.cls]: Retrieved the Opportunity based on the OCR's Opportunity Id and compared the Account.

File: src/classes/OPP_AutomatedSoftCreditsService_TEST.cls [hunk 4]
```diff
@@ -53,15 +54,26 @@ private class OPP_AutomatedSoftCreditsService_TEST {
     * @return void
     **********************************************************************************************************/
     @testSetup static void createAutomatedSoftCreditsData() {
-        List<Contact> testContacts = UTIL_UnitTestData_TEST.CreateMultipleTestContacts(5);
+        Account acct1 = new Account(Name = 'Organization Account1');
+        Account acct2 = new Account(Name = 'Organization Account2');
+        acct1.npe01__SYSTEMIsIndividual__c = false;
```
[the remark this answers is not in the document]
Comment (joy-sfdo) [reply src/classes/OPP_AutomatedSoftCreditsService_TEST.cls]: Removed the code that sets the npe01__SYSTEMIsIndividual__c flag.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls (hunk 2 above)
Comment (joy-sfdo) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: Added the usage of the isOrganizationalAccount method.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 5]
```diff
@@ -289,11 +292,19 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
...
 
+        // Create Relationship Soft Credits accordingly
         if (!primaryContactsToOpportunities.isEmpty()) {
             List<sObject> relationshipOCRs
                 = OPP_AutomatedSoftCreditsService.createRelationshipOCRs(primaryContactsToOpportunities);
             dmlWrapper.objectsToInsert.addAll(relationshipOCRs);
         }
+
+        // Create Affiliation Soft Credits accordingly
+        if (!accountsToOpportunities.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (joy-sfdo) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: During testing, did you come across a scenario that created duplicate OCRs? The Relationship OCRs are created for Opportunities that have an Individual Account; whereas, the Affiliation OCRs are created for Opportunities that have an Organization Account. Given that, the OCRs created would not be duplicated because of the differing Opportunity records. Please share how you were able to create duplicates. I'd like to add it to our tests.

There is the potential for OPP_AutomatedSoftCreditsService.retrieveOpportunityContactRoles() to be called twice. However, unless there is a bulk process with both Individual and Organization Account Opportunities, then, generally, only one call will be made to that method. Mapping Automated Soft Credits would provide insights about its expected usage.  

I agree there are alternative ways to implement this code. In fact, it would have been great to refactor the OPP_OpportunityContactRoles_TDTM class in general, but I know the preference has been to make minimal changes to the code. If you think the OPP_OpportunityContactRoles_TDTM class now needs to be refactored, let's discuss with Mary and Alex.

**Comment by davidhabib** [comment]
So here are some issues I've run into:

1. if I have a Relationship between two household members, and the relationships specify a contact role, then when an Opp is saved with one of the contacts being the primary contact, there is the correct OCR from the relationship created.  But the 'Household Member' OCR, that is normally created, is missing.

2. If I have two household members, with Relationships between them specifying OCR role 'Soft Credit', and both contacts also have an Affiliation to an organizational account (such as their foundation), specifying an OCR role 'Foundation Donor'.  Then when I create an organizational donation from their foundation account, and specify one of the contacts as the primary donor, I do get OCR's for both with 'Foundation Donor' and the correct contact's OCR is marked Primary.  But I do not get OCR's for Soft Credit.  I assume this is because your code is only applying Relationship OCR's for Individual donations, not Org donations.  We need to get agreement with @mpusto on whether this is the correct behavior or not.

3. I have two household members with an Organization account and Affiliations to that account.  The affliations have OCR role 'Soft Credit'.  The NPSP Settings for Default Organizational Contact Role is set to 'Matched Donor'.  Now create an organizational Opportunity with one of the contacts as primary contact.  You end up with 3 OCR's (Soft Credit, Soft Credit, Matched Donor) which is correct, but both of the OCR's for the primary contact is marked Primary, which is incorrect.  only one should be marked primary, and I believe it should be the Matched Donor one.

Rather than continuing more adhoc testing that I think might find more issues, let's look at this thoroughly.  There are 4 ways that OCR's get auto created in NPSP:

* the OCR for the Primary Contact on the Opportunity (controlled by the Default Opportunity Contact Role and Default Organizational Contact Role NPSP Settings)
* the household member OCR's (controlled by the Household Member Contact Role and Household Contact Roles On NPSP Settings)
* the Relationships OCR's
* the Affiliation OCR's

I think we need to test the interplay of all these settings with both individual Opps and org Opps, to ensure the correct OCR's result.  I bet we could come up with a single helper test routine that took flags to specify which of the OCR Settings to consider, and what the expected results to be.  Then you could easily test the full set of permutations.  Creating the matrix of expected results would also help us ensure we have the full spec thought out.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidhabib) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: You may be right that we should only do Rels for individual opps, and Affls for organizational opps.  I couldn't remember if that was discussed and resolved or not.  When we discuss with @mpusto make sure to bring up the scenario of a contact who uses their business or foundation account for donations, but may have relationships to a board member who should get soft credited.

I can't say what the frequency of both org opps and individual opps being inserted at the same time.  It may be rare, but in general I just always try to design code optimized for bulk operations assuming a mix.

You are right that we do try to keep minimal changes to existing code to avoid breaking existing functionality.  that said, I'd still want us to consider having the old code use a new shared cache of OCR's to avoid multiple soql calls.  As for doing a complete redesign, I'd be less excited about going that route, unless we found that there are additional issues uncovered in this trigger handler that would be better resolved with more of a redesign.

[the remark this answers is not in the document]
Comment (joy-sfdo) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: Sounds good. Let's follow up with @mpusto.

**Comment by joy-sfdo** [comment]
@davidhabib -- Thanks for the scenarios. The ad hoc/edge case testing is helpful in presenting the possible usage of the feature. Currently, we have in place test charters for three of the four scenarios listed:
- the OCR for the Primary Contact on the Opportunity (controlled by the Default Opportunity Contact Role and Default Organizational Contact Role NPSP Settings)  -- The focus has been on the new Organizational setting. 
- the Relationships OCRs
- the Affiliation OCRs

I'll follow up with @mpusto and the team about the Household member scenarios as well as mapping out the feature's usage, which is basically creating a matrix. I really like the mapping/matrix to test the interplay of Individual and Organizational Opportunities and their OCRs.

**Comment by joy-sfdo** [comment]
Keeping everyone in the loop regarding the issues Dave raised on Monday, April 23rd.

Scenario #1 -- I'm currently working on the fix for this issue. Extra time is required to review the code, implement the change, and test the fix in order to ensure that other parts of the code aren't impacted because the issue is inside of a single method that performs multiple tasks.

Scenario #2 -- During a discussion with the team on April 25th, it was decided by @mpusto that Opportunity Contact Role records will not be created for the Primary Contact's Relationships on an Organizational Gift at this time.

Scenario #3 -- The issue described reflects salesforce standard behavior. The IsPrimary field is "Defaulted on create" and per the documentation (https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_opportunitycontactrole.htm): 
Indicates whether the associated Contact plays the primary role on the Opportunity (true) or not (false). Each Opportunity has only one primary contact. Label is Primary.

Please reach out with questions.

@alex-peralta, @force2b

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 6]
```diff
@@ -654,4 +677,34 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
...
+    private static Map<Id, List<Id>> populateAccountToOpportunities(List<Opportunity> currentOpportunities, Map<Id, Account> oppAccountDetails) {
+        Map<Id, List<Id>> accountToOpportunity = new Map<Id, List<Id>>();
+
+        if (currentOpportunities.isEmpty()) {
+            return accountToOpportunity;
+        }
+
+        for (Opportunity eachOppty : currentOpportunities) {
+            if (eachOppty.AccountId != null
+                && OPP_AutomatedSoftCreditsService.isOrganizationalAccount(oppAccountDetails.get(eachOppty.AccountId).npe01__SYSTEMIsIndividual__c)) {
```
[the remark this answers is not in the document]
Comment (joy-sfdo) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: Thanks for the input! I'd be happy to chat about this method.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls [hunk 7]
```diff
@@ -654,4 +677,34 @@ public without sharing class OPP_OpportunityContactRoles_TDTM extends TDTM_Runna
...
 
+    /*********************************************************************************************************
+    * @description Populates a Map of Account Id to Opportunity Ids.
+    * @param currentOpportunities The Opportunities in the 'new' trigger context variable.
+    * @param oppAccountDetails A Map containing the Opportunities' related Accounts.
+    * @return Map<Id, Id> contains associated Primary Contacts and Opportunity Ids.
+    **********************************************************************************************************/
+    @testVisible
+    private static Map<Id, List<Id>> populateAccountToOpportunities(List<Opportunity> currentOpportunities, Map<Id, Account> oppAccountDetails) {
+        Map<Id, List<Id>> accountToOpportunity = new Map<Id, List<Id>>();
```
[the remark this answers is not in the document]
Comment (joy-sfdo) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: I'm not a fan of adding the data structure to the variable name. The method is small enough for someone to see it's a List of Ids. With that being said, accountIdToOpportunityIds could work.

File: src/classes/OPP_OpportunityContactRoles_TEST.cls [hunk 8]
```diff
@@ -713,4 +717,40 @@ private class OPP_OpportunityContactRoles_TEST {
...
 
 
+    /*********************************************************************************************************
+    * @description Tests populating a Map of the Opportunity's associated Account Id to  Opportunity Ids.
+    * verify:
+    *   The Map was populated accordingly.
+    * @return void
+    *********************************************************************************************************/
+    static testMethod void testPopulateAccountToOpportunities() {
+        Map<Id, List<Id>> accountToOpportunity = new Map<Id, List<Id>>();
```
[the remark this answers is not in the document]
Comment (joy-sfdo) [reply src/classes/OPP_OpportunityContactRoles_TEST.cls]: Same comment as above.

**Comment by joy-sfdo** [comment]
Thanks so much for jumping in and testing the code, Mike! 

I'll schedule some time for us to chat about the items called out.

File: src/classes/OPP_OpportunityContactRoles_TDTM.cls (hunk 7 above)
[the remark this answers is not in the document]
Comment (joy-sfdo) [reply src/classes/OPP_OpportunityContactRoles_TDTM.cls]: I just updated the variable names for both the accountToOpportunity and primaryContactToOpportunity variables to match the method signatures.

Good catch! Thanks!
